### ISSUE TITLE
no-unused-vars false positive for exported function

### ISSUE THREAD

**Issue by silviujingoi** [eslint/eslint#12573]
**Tell us about your environment**

* **ESLint Version:** 6.3.0
* **Node Version:** v8.11.2
* **npm Version:** 5.6.0

**What parser (default, Babel-ESLint, etc.) are you using?** @typescript-eslint/parser

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
module.exports = {
  env: {
    browser: true,
    es6: true,
  },
  extends: [
    'eslint:recommended',
    'plugin:prettier/recommended',
  ],
  "globals": {
    "$": false
  },
  "rules": {
  }
};
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
/* exported myFunc */
/* exported myVar */
{
  var myVar = 1;

  function myFunc() {
    return 1;
  }
}
```

```bash
./node_modules/.bin/eslint path/to/file
```

**What did you expect to happen?**

No error for myFunc (just like for myVar)

**What actually happened? Please include the actual, raw output from ESLint.**

   69:12  error  'myFunc' is defined but never used          no-unused-vars

**Are you willing to submit a pull request to fix this bug?**

**Comment by kaicataldo** [comment]
I believe this is working as intended, since you are using ES2015 modules and `myFunc` is block scoped and therefore can't be exported. Note that `var` is not block scoped (if you change it to `let` or `const` you would see the same error. Example [here](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXhwb3J0ZWQgbXlGdW5jICovXG4vKiBleHBvcnRlZCBteVZhciAqL1xuLyogZXhwb3J0ZWQgbXlWYXIyICovXG4vKiBleHBvcnRlZCBteVZhcjIgKi9cblxue1xuICB2YXIgbXlWYXIgPSAxO1xuICBsZXQgbXlWYXIyID0gMTtcbiAgY29uc3QgbXlWYXIzID0gMTtcblxuICBmdW5jdGlvbiBteUZ1bmMoKSB7XG4gICAgcmV0dXJuIDE7XG4gIH1cbn0iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjExLCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7ImNvbnN0cnVjdG9yLXN1cGVyIjoyLCJmb3ItZGlyZWN0aW9uIjoyLCJnZXR0ZXItcmV0dXJuIjoyLCJuby1hc3luYy1wcm9taXNlLWV4ZWN1dG9yIjoyLCJuby1jYXNlLWRlY2xhcmF0aW9ucyI6Miwibm8tY2xhc3MtYXNzaWduIjoyLCJuby1jb21wYXJlLW5lZy16ZXJvIjoyLCJuby1jb25kLWFzc2lnbiI6Miwibm8tY29uc3QtYXNzaWduIjoyLCJuby1jb25zdGFudC1jb25kaXRpb24iOjIsIm5vLWNvbnRyb2wtcmVnZXgiOjIsIm5vLWRlYnVnZ2VyIjoyLCJuby1kZWxldGUtdmFyIjoyLCJuby1kdXBlLWFyZ3MiOjIsIm5vLWR1cGUtY2xhc3MtbWVtYmVycyI6Miwibm8tZHVwZS1rZXlzIjoyLCJuby1kdXBsaWNhdGUtY2FzZSI6Miwibm8tZW1wdHkiOjIsIm5vLWVtcHR5LWNoYXJhY3Rlci1jbGFzcyI6Miwibm8tZW1wdHktcGF0dGVybiI6Miwibm8tZXgtYXNzaWduIjoyLCJuby1leHRyYS1ib29sZWFuLWNhc3QiOjIsIm5vLWV4dHJhLXNlbWkiOjIsIm5vLWZhbGx0aHJvdWdoIjoyLCJuby1mdW5jLWFzc2lnbiI6Miwibm8tZ2xvYmFsLWFzc2lnbiI6Miwibm8taW5uZXItZGVjbGFyYXRpb25zIjoyLCJuby1pbnZhbGlkLXJlZ2V4cCI6Miwibm8taXJyZWd1bGFyLXdoaXRlc3BhY2UiOjIsIm5vLW1pc2xlYWRpbmctY2hhcmFjdGVyLWNsYXNzIjoyLCJuby1taXhlZC1zcGFjZXMtYW5kLXRhYnMiOjIsIm5vLW5ldy1zeW1ib2wiOjIsIm5vLW9iai1jYWxscyI6Miwibm8tb2N0YWwiOjIsIm5vLXByb3RvdHlwZS1idWlsdGlucyI6Miwibm8tcmVkZWNsYXJlIjoyLCJuby1yZWdleC1zcGFjZXMiOjIsIm5vLXNlbGYtYXNzaWduIjoyLCJuby1zaGFkb3ctcmVzdHJpY3RlZC1uYW1lcyI6Miwibm8tc3BhcnNlLWFycmF5cyI6Miwibm8tdGhpcy1iZWZvcmUtc3VwZXIiOjIsIm5vLXVuZGVmIjoyLCJuby11bmV4cGVjdGVkLW11bHRpbGluZSI6Miwibm8tdW5yZWFjaGFibGUiOjIsIm5vLXVuc2FmZS1maW5hbGx5IjoyLCJuby11bnNhZmUtbmVnYXRpb24iOjIsIm5vLXVudXNlZC1sYWJlbHMiOjIsIm5vLXVudXNlZC12YXJzIjoyLCJuby11c2VsZXNzLWNhdGNoIjoyLCJuby11c2VsZXNzLWVzY2FwZSI6Miwibm8td2l0aCI6MiwicmVxdWlyZS1hdG9taWMtdXBkYXRlcyI6MiwicmVxdWlyZS15aWVsZCI6MiwidXNlLWlzbmFuIjoyLCJ2YWxpZC10eXBlb2YiOjIsInF1b3RlcyI6Miwibm8tdGVtcGxhdGUtY3VybHktaW4tc3RyaW5nIjoyfSwiZW52Ijp7fX19).

I'm a bit confused though - are you actually using modules? The `exported` comment is only necessary when you're using scripts that define things on the global object.

**Comment by mdjermanovic** [comment]
It would be a bug if the rule doesn't report `var` as well when source type is set to `module`?

But it seems to work well in [this demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXhwb3J0ZWQgbXlGdW5jICovXG4vKiBleHBvcnRlZCBteVZhciAqL1xue1xuICB2YXIgbXlWYXIgPSAxO1xuXG4gIGZ1bmN0aW9uIG15RnVuYygpIHtcbiAgICByZXR1cm4gMTtcbiAgfVxufSIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6Niwic291cmNlVHlwZSI6Im1vZHVsZSIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6eyJuby11bnVzZWQtdmFycyI6Mn0sImVudiI6e319fQ==), both `myVar` and `myFunc` are reported.

**Comment by silviujingoi** [comment]
> I believe this is working as intended, since you are using ES2015 modules
> I'm a bit confused though - are you actually using modules? The `exported` comment is only necessary when you're using scripts

My apologies, I posted the wrong configuration file.  I updated the original post.  I am not using modules for this project.

**Comment by mdjermanovic** [comment]
ESLint's scope manager treats function declarations as block-scoped in ES2015+, even in non-strict mode. Therefore, the rule doesn't see `myFunc` as being exported.

It doesn't work like that in browsers because they are required to support legacy code. The example from the original post would indeed create global `myFunc` (in non-strict mode only).

But, it's a legacy feature that should be avoided in new code, e.g. by using `var myFunc = function () {}` or `window.myFunc = function () {}`
